common/isolationgroup/default-isolation-group-state.go

```diff
@@ -0,0 +1,161 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package isolationgroup
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/common/types"
+)
+
+type defaultIsolationGroupStateHandler struct {
+	status                     int32
+	done                       chan bool
+	log                        log.Logger
+	domainCache                cache.DomainCache
+	globalIsolationGroupDrains persistence.ConfigStore
+	config                     Config
+	subscriptionMu             sync.Mutex
+	valuesMu                   sync.RWMutex
+	lastSeen                   *IsolationGroups
+	updateCB                   func()
+	// subscriptions is a map of domains->subscription-keys-> subscription channels
+	// for notifying when there's a state change
+	subscriptions map[string]map[string]chan<- ChangeEvent
+}
+
+func NewDefaultIsolationGroupStateWatcher(
+	logger log.Logger,
+	domainCache cache.DomainCache,
+	config Config,
+	globalIsolationGroupDrains persistence.ConfigStore,
+	done chan bool,
+) State {
+	return &defaultIsolationGroupStateHandler{
+		status:                     common.DaemonStatusInitialized,
+		log:                        logger,
+		config:                     config,
+		domainCache:                domainCache,
+		globalIsolationGroupDrains: globalIsolationGroupDrains,
+		subscriptionMu:             sync.Mutex{},
+		subscriptions:              make(map[string]map[string]chan<- ChangeEvent),
+	}
+}
+
+func (z *defaultIsolationGroupStateHandler) GetByDomainID(ctx context.Context, domainID string) (*IsolationGroups, error) {
+	domain, err := z.domainCache.GetDomainByID(domainID)
+	if err != nil {
+		return nil, fmt.Errorf("could not resolve domain in isolationGroup handler: %w", err)
+	}
+	return z.Get(ctx, domain.GetInfo().Name)
+}
+
+// Get the statue of a isolationGroup, with respect to both domain and global drains. Domain-specific drains override global config
+// will return nil, nil when it is not enabled
+func (z *defaultIsolationGroupStateHandler) Get(ctx context.Context, domain string) (*IsolationGroups, error) {
+	if !z.config.zonalPartitioningEnabledForDomain(domain) {
+		return nil, nil
+	}
+
+	domainData, err := z.domainCache.GetDomain(domain)
+	if err != nil {
+		return nil, fmt.Errorf("could not resolve domain in isolationGroup handler: %w", err)
+	}
+	domainState := domainData.GetInfo().IsolationGroupConfig
+	globalCfg, err := z.globalIsolationGroupDrains.FetchConfig(ctx, persistence.GlobalIsolationGroupConfig)
```

Comment:
I think we need some cache for `configStore`

Reply:
Yes, but if it's ok let me follow it up. I'd like to do that in a subsequent PR